## Laboratorio 4 realizado por: Johan Suarez y Kalia Gonzalez

### Ejercicio 1 

A continuacion , se quiere realizar el siguiente ejericio usando proramacion lineal con el metodo Simplex estandarizado: $$\text{max}(3x1+2x2+5x3)$$ y donde esta sujeto a las siguientes instrucciones: $$\text(x1+x2+x3\leq 100)$$ $$\text(2x1+x2+x3 \leq 150)$$ $$\text(x1+4x2+2x3 \leq 80)$$  $$\text(x1,x2,x3 \geq 0)$$


##### Explicacion de Implementacion

1. Agregamos las variables de holgura s1,s2 y s3 e igualamos las ecuaciones. 
2. Iteramos con base a la formula matematica $$\text (cj = cj-\lambda Aj)$$ done cj representa la matriz de coeficientes de cada ecuacion incluyendo el de las variables. 
3. Iteramos hasta encontrar el optimo

In [10]:
import numpy as np 

#Definimos una lista donde estaran los coeficientes de cada una de las variables incluyendo las de holgura 
coeficiente =np.array([-3,-2,-5,0,0,0])# Ponemos las variables de holgura como 0 y las variables de decision como los coeficientes de la funcion objetivo en negativo 
# Definimos la matriz de restricciones 
restricciones = np.array([[1,1,1,1,0,0],[2,1,1,0,1,0],[1,4,2,0,0,1]]) 
# Definimos el vector de resultados 
resultados = np.array([100,150,80]) 
#Definimos las variables de holgura como base (columna 3,4,5)
variables_base = np.array([3,4,5]) 
#Realizamos el algoritmo 
def simplexAlgorithm(coeficientes,restricciones,resultados): 
    m,n = restricciones.shape  
    # Definimos la matriz de coeficientes
    b = restricciones[:, variables_base] 
    #Lista de los indices de las variables no basicas 
    non_basic = [j for j in range(n) if j not in variables_base]
    # Definimos la matriz de soluciones 
    while True: 
        b_invertida = np.linalg.inv(b) 
        x_b = (b_invertida @ resultados ).flatten()
        lambda_ = coeficientes[variables_base] @ b_invertida 
        #Calculamos el costo reducido 
        costo_reducido = coeficientes[non_basic] - lambda_ @ restricciones[:, non_basic] 
        # Verificamos si el costo reducido es positivo 
        if all(costo_reducido >= 0): 
            x = np.zeros(n) 
            for i , var_index in enumerate(variables_base): 
                x[var_index] =float(x_b[i] )
            z = -coeficientes @ x 
            return x, z 
        # Encontramos la variable entrante 
        entrante = non_basic[np.argmin(costo_reducido)] 
        d = b_invertida @ restricciones[:, entrante] 
        ratios = x_b / d 
        #Determinamos cual varibale basica va a salir de la base
        fuera_indice = np.argmin(ratios) 
        fuera = variables_base[fuera_indice] 
        # Actualizamos la base 
        variables_base[fuera_indice] = entrante 
        b = restricciones[:, variables_base] 
        non_basic = [ j for j in range(n) if j not in variables_base] 
#Ejecutamos el algoritmo 
solucion , z = simplexAlgorithm(coeficiente, restricciones, resultados) 
print("La solucion es: ",solucion) 
print("El valor de la funcion objetivo es: ",z)
        

La solucion es:  [73.33333333  0.          3.33333333 23.33333333  0.          0.        ]
El valor de la funcion objetivo es:  236.66666666666666


#### Tablas de iteraciones

A continuacion se muestran las tablas de cada iteracion: 

| Basica | x1 | x2 | x3 | s1 | s2 | s3 | RHS | 
|--------|----|----|----|----|----|----|-----| 
| s1     | 1  | 1  | 1  | 1  | 0  | 0  | 100 | 
| s2     | 2  | 1  | 1  | 0  | 1  | 0  | 150 | 
| s3     | 1  | 4  | 2  | 0  | 0  | 1  | 80  |  
| zj-cj     | -3  | -2  | -5  | 0  | 0  | 0  |  | 




Ahora buscamos la proxima variable a introducir en la base. En este caso es -5 en la columna 3.  Dividimos cada termino de dicha columna con el resultado (RHS) : $$\text(100/1 = 150)$$ $$\text(150/1 = 150)$$ $$\text(80/2 = 40)$$

| Basica | x1 | x2 | x3 | s1 | s2 | s3 | RHS | 
|--------|----|----|----|----|----|----|-----| 
| s1     | 0.5  | -1  | 0 | 1  | 0  | 0  | 60 | 
| s2     | 1.5  | -1  | 0  | 0  | 1  | 0  | 110 | 
| s3     | 0.5  | 2  | 1 | 0  | 0  | 0.5  | 40  |  
| zj-cj     | -0.5  | 8  | 0  | 0  | 0  | 0  |  | 

Realizamos la ultima iteracion dividiendo los resultados con los valores en la columna 2 (x1) :  $$\text(60/0.5 = 120)$$ $$\text(110/1.5= 73,33)$$ $$\text(40/0.5 = 80)$$

| Basica | x1 | x2 | x3 | s1 | s2 | s3 | RHS | 
|--------|----|----|----|----|----|----|-----| 
| s1     | 0  | -0.67  | 0 | 1  | -0.33  | -0.33  | 23.33 | 
| s2     | 1 | -0.67 | 0  | 0  | -0.67  | -0.33  | 73.33 | 
| s3     | 0  | 2.33  | 1 | 0  | -0.33  | 0.67  | 3.33  |  
| zj-cj     | -0 | 7.67  | 0  | 0  | -0.33  | 2.33  |  | 

#### Analisis Geometrico: 

Este tipo de problema define una region que esta acostada por planos correspondientes a cada restriccion. Cada una de ellas delimita una parte del espacio. Esto confirma que el optimo se alcanza en la interseccion de las dos restricciones activas (mas limitantes). HACER GRAFICASSSSSS

#### Analisis de Sensibilidad : 

In [11]:
import numpy as np 

#Definimos una lista donde estaran los coeficientes de cada una de las variables incluyendo las de holgura 
coeficiente =np.array([-3,-2,-5,2.33,0,0])# Ponemos las variables de holgura como 0 y las variables de decision como los coeficientes de la funcion objetivo en negativo 
# Definimos la matriz de restricciones 
restricciones = np.array([[1,1,1,1,0,0],[2,1,1,0,1,0],[1,4,2,0,0,1]]) 
# Definimos el vector de resultados 
resultados = np.array([100,150,80]) 
#Definimos las variables de holgura como base (columna 3,4,5)
variables_base = np.array([3,4,5]) 
#Realizamos el algoritmo 
def simplexAlgorithm(coeficientes,restricciones,resultados): 
    m,n = restricciones.shape  
    # Definimos la matriz de coeficientes
    b = restricciones[:, variables_base] 
    #Lista de los indices de las variables no basicas 
    non_basic = [j for j in range(n) if j not in variables_base]
    # Definimos la matriz de soluciones 
    while True: 
        b_invertida = np.linalg.inv(b) 
        x_b = (b_invertida @ resultados ).flatten()
        lambda_ = coeficientes[variables_base] @ b_invertida 
        #Calculamos el costo reducido 
        costo_reducido = coeficientes[non_basic] - lambda_ @ restricciones[:, non_basic] 
        # Verificamos si el costo reducido es positivo 
        if all(costo_reducido >= 0): 
            x = np.zeros(n) 
            for i , var_index in enumerate(variables_base): 
                x[var_index] =float(x_b[i] )
            z = -coeficientes @ x 
            return x, z 
        # Encontramos la variable entrante 
        entrante = non_basic[np.argmin(costo_reducido)] 
        d = b_invertida @ restricciones[:, entrante] 
        ratios = x_b / d 
        #Determinamos cual varibale basica va a salir de la base
        fuera_indice = np.argmin(ratios) 
        fuera = variables_base[fuera_indice] 
        # Actualizamos la base 
        variables_base[fuera_indice] = entrante 
        b = restricciones[:, variables_base] 
        non_basic = [ j for j in range(n) if j not in variables_base] 
#Ejecutamos el algoritmo 
solucion , z = simplexAlgorithm(coeficiente, restricciones, resultados) 
print("La solucion es: ",solucion) 
print("El valor de la funcion objetivo es: ",z)
        

La solucion es:  [73.33333333  0.          3.33333333 23.33333333  0.          0.        ]
El valor de la funcion objetivo es:  182.29999999999998


En este caso, vemos que al cambiar los valores de algunos de las variables de holgura , el resultado cambia drasticamente y resulta ser menor al que es originalmente. Esto tal vez se deba por la nueva asignacion de pibotes durante el calculo de las variables que hagan parte o no de la matriz.

In [13]:
import numpy as np 

#Definimos una lista donde estaran los coeficientes de cada una de las variables incluyendo las de holgura 
coeficiente =np.array([-3,12,-5,0,0,0])# Ponemos las variables de holgura como 0 y las variables de decision como los coeficientes de la funcion objetivo en negativo 
# Definimos la matriz de restricciones 
restricciones = np.array([[1,1,1,1,0,0],[2,1,1,0,1,0],[1,4,2,0,0,1]]) 
# Definimos el vector de resultados 
resultados = np.array([100,150,80]) 
#Definimos las variables de holgura como base (columna 3,4,5)
variables_base = np.array([3,4,5]) 
#Realizamos el algoritmo 
def simplexAlgorithm(coeficientes,restricciones,resultados): 
    m,n = restricciones.shape  
    # Definimos la matriz de coeficientes
    b = restricciones[:, variables_base] 
    #Lista de los indices de las variables no basicas 
    non_basic = [j for j in range(n) if j not in variables_base]
    # Definimos la matriz de soluciones 
    while True: 
        b_invertida = np.linalg.inv(b) 
        x_b = (b_invertida @ resultados ).flatten()
        lambda_ = coeficientes[variables_base] @ b_invertida 
        #Calculamos el costo reducido 
        costo_reducido = coeficientes[non_basic] - lambda_ @ restricciones[:, non_basic] 
        # Verificamos si el costo reducido es positivo 
        if all(costo_reducido >= 0): 
            x = np.zeros(n) 
            for i , var_index in enumerate(variables_base): 
                x[var_index] =float(x_b[i] )
            z = -coeficientes @ x 
            return x, z 
        # Encontramos la variable entrante 
        entrante = non_basic[np.argmin(costo_reducido)] 
        d = b_invertida @ restricciones[:, entrante] 
        ratios = x_b / d 
        #Determinamos cual varibale basica va a salir de la base
        fuera_indice = np.argmin(ratios) 
        fuera = variables_base[fuera_indice] 
        # Actualizamos la base 
        variables_base[fuera_indice] = entrante 
        b = restricciones[:, variables_base] 
        non_basic = [ j for j in range(n) if j not in variables_base] 
#Ejecutamos el algoritmo 
solucion , z = simplexAlgorithm(coeficiente, restricciones, resultados) 
print("La solucion es: ",solucion) 
print("El valor de la funcion objetivo es: ",z)
        

La solucion es:  [73.33333333  0.          3.33333333 23.33333333  0.          0.        ]
El valor de la funcion objetivo es:  236.66666666666666


Sin embargo, en este otro caso a pesar de haber cambiado el valor de la variable objetivo, nos da el mismo resultado. Convirtiendo asi la solucion estable ante distintos cambios.